aten/src/ATen/new_gen.py

```diff
@@ -0,0 +1,953 @@
+        # We should probably get a proper parser here
+        assert ' -> ' in func, "function schema missing return type (spaces are mandatory)"
+        func_decl, return_decl = [x.strip() for x in func.split(' -> ')]
+        ops, args = func_decl.split('(', 1)
+        assert args[-1] == ")", "Expecting closing )"
+        args = args[:-1]
+        name = OperatorName.parse(ops)
+        arguments, kwarg_only_arguments, out_arguments = parse_arguments(args)
+        returns = parse_return_arguments(return_decl)
+        r = FunctionSchema(
+            name=name,
+            arguments=arguments,
+            kwarg_only_arguments=kwarg_only_arguments,
+            out_arguments=out_arguments,
+            returns=returns
+        )
+        assert str(r) == func, f'{str(r)} != {func}'
+        return r
+
+    def __post_init__(self) -> None:
+        for arg, ret in zip(self.out_arguments, self.returns):
+            assert arg.annotation == ret.annotation, \
+                "Out arguments must have matching return Tensor; furthermore, " \
+                "the ith-argument needs to correspond to the ith return"
+        if self.out_arguments:
+            assert len(self.out_arguments) == len(self.returns), \
+                "Must return as many arguments as there are out arguments"
+        if self.name.name.inplace:
+            # TODO: fixme
+            if str(self.name) not in ['_amp_non_finite_check_and_unscale_']:
+                assert len(self.returns) == 1
+
+    def is_out_fn(self) -> bool:
+        # Note [is_out_fn]
+        #
+        # out functions are the variants which take an explicit out= argument
+        # to populate into.  We need to know if a schema corresponds to an
+        # out function for several reasons:
+        #
+        #   - They codegen differently in C++ API
+        #       - codegen to at::add_out rather than at::add
+        #       - out argument is moved to front of C++ argument list
+        #
+        # out functions are DEFINED to be any function with a keyword-only
+        # argument that is mutable.  In principle, this could lead to a
+        # false positive if you define a function that mutates a
+        # kwarg only argument, but this isn't the "true" output of this
+        # function.  A more robust definition that would work in this
+        # case would also look at:
+        #
+        #   - The output types.  Out functions take in the arguments
+        #     they mutate and then return them again; this is sort
+        #     of "definitionally" what makes something an out function.
+        #     Historically, we DO check this for consistency.
+        #   - Correspondence with pure variant.  An out function
+        #     should have a signature equivalent to its pure variant,
+        #     but just with extra kwargs for the output elements.  This
+        #     is difficult to actually check for and historically
+        #     we only do this check in tools/
```

Comment:
This is the kind of entropy reduction we'd harvest by merging (or sharing utilities across) ATen and tools/ codegen: there's a bunch of logic over there that does semantic checking over the same interpretations of these definitions. Ideally we'd eventually merge that into what you're doing here and use that canonical model everywhere.

Reply:
cc @ljk53 per our recent convos around codegen

Reply:
E.g. [here](https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_python_functions.py#L1300-L1375)'s where the tools/ side carefully undoes everything :)